File: engine_py/tests/test_pathfinding.py
def test_pathfinding(make_world):
    world = make_world()
    for x in range(3):
        for y in range(3):
            world.add_cell(x, y, 0)
    for x in range(3):
        for y in range(3):
            for dx, dy in [(1,0),(0,1),(-1,0),(0,-1)]:
                nx, ny = x+dx, y+dy
                if 0<=nx<=2 and 0<=ny<=2:
                    world.add_neighbor((x,y,0), (nx,ny,0))
    world.set_cell_metadata({"Square": {"x":1,"y":1,"z":0}}, {"walkable": False})
    result = world.find_path({"Square": {"x":0,"y":0,"z":0}}, {"Square": {"x":2,"y":2,"z":0}})
    assert result is not None
    path = result["path"]
    for cell in path:
        if "Square" in cell:
            assert not (cell["Square"]["x"]==1 and cell["Square"]["y"]==1)
    assert len(path) == 5

File: engine/scripts/lua/tests/test_pathfinding.lua
local assert = require("assert")

local function test_pathfinding()
	-- Build a 3x3 grid
	for x = 0, 2 do
		for y = 0, 2 do
			add_cell(x, y, 0)
		end
	end
	for x = 0, 2 do
		for y = 0, 2 do
			for _, d in ipairs({ { 1, 0 }, { 0, 1 }, { -1, 0 }, { 0, -1 } }) do
				local nx, ny = x + d[1], y + d[2]
				if nx >= 0 and nx <= 2 and ny >= 0 and ny <= 2 then
					add_neighbor({ x = x, y = y, z = 0 }, { x = nx, y = ny, z = 0 })
				end
			end
		end
	end
	-- Block (1,1,0)
	set_cell_metadata({ x = 1, y = 1, z = 0 }, { walkable = false })
	local result = find_path({ x = 0, y = 0, z = 0 }, { x = 2, y = 2, z = 0 })
	assert.is_table(result)
	assert.is_table(result.path)
	for _, cell in ipairs(result.path) do
		assert.is_false(cell.x == 1 and cell.y == 1)
	end
	assert.equals(#result.path, 5)
end

return { test_pathfinding = test_pathfinding }
